Add functionality to add restaurants to LikeList component

evan-m-jackson opened the pull request:
- Add `likedRestaurant` variable that is an array of Restaurant objects that are liked by the user
- Add `addLikedRestaurant` function to updated `likedRestaurant` variable with new liked Restaurants
- Rename `MatchList` component to `LikedList`

evan-m-jackson committed:
Add Liked Restaurants to LikeList component

- Add `likedRestaurant` variable that is an array of Restaurant objects that are liked by the user
- Add `addLikedRestaurant` function to updated `likedRestaurant` variable with new liked Restaurants
- Rename `MatchList` component to `LikedList`

clairenguyen commented on frontend/__tests__/components/CardCarousel.test.tsx:
```diff
@@ -34,21 +34,25 @@ describe('CardCarousel', () => {
   })
+
+  test('clicking on accept button calls the liked function', async () => {
+    const likeRestaurant = jest.fn()
```
nitpicking:  I'd name this something to be more consistent with how it's used. e.g. `addLikedRestaurantMock`.

clairenguyen commented on frontend/__tests__/components/CardScreen.test.tsx:
```diff
@@ -2,52 +2,54 @@ import { act, render, screen } from '@testing-library/react'
 
     await act(() => {
-      render(<CardScreen />)
+      render(<CardScreen addLikedRestaurant={likedRestaurant} />)
```
nitpick (non-blocking): for the next two tests, I personally wouldn't bother to even create a variable for this mocked prop.  I would just do `<CardScreen addLikedRestaurant={jest.fn()} />` and inline the mock. This communicates that it doesn't matter what the prop is as it doesn't affect the test.

clairenguyen commented on frontend/__tests__/components/LikeList.test.tsx:
```diff
@@ -0,0 +1,61 @@
+  website_url: 'http://www.dummyrestaurant.com',
+}
+
+const LIKED_RESTAURANTS = [FAKE_RESTAURANT_ONE, FAKE_RESTAURANT_TWO, FAKE_RESTAURANT_THREE]
```
@seoulection  made a good point on this PR https://github.com/8l-project-jam/tastemate/pull/55#discussion_r1358569798 that the all caps variables are typically for variables that are globally scoped. Can we change these to lower case?

clairenguyen commented on frontend/__tests__/components/LikeList.test.tsx:
```diff
@@ -0,0 +1,61 @@
+const LIKED_RESTAURANTS = [FAKE_RESTAURANT_ONE, FAKE_RESTAURANT_TWO, FAKE_RESTAURANT_THREE]
+
+describe('LikeList', () => {
+  it('renders the TasteMate h1', () => {
```
For test names, we should try to avoid mentioning things like h1 as that's an implementation detailed.  Maybe instead, just say, `displays the TasteMate header`?

clairenguyen commented on frontend/__tests__/components/LikeList.test.tsx:
```diff
@@ -0,0 +1,61 @@
+describe('LikeList', () => {
+  it('renders the TasteMate h1', () => {
+    render(<LikeList likedRestaurants={[]} />)
+    const myElem = screen.getByText('TasteMate')
```
Can we rename this variable? `myElem` doesn't really communicate much to me. I'd rather something like `header`.

clairenguyen commented on frontend/__tests__/components/LikeList.test.tsx:
```diff
@@ -0,0 +1,61 @@
+    render(<LikeList likedRestaurants={[]} />)
+    const myElem = screen.getByText('TasteMate')
+
+    expect(myElem).toBeInTheDocument()
```
nitpick:  

Generally, React Testing Library emphasizes writing tests that align closely with user interactions and experience, ie. testing what the user actually sees and how they may interact with the UI.

It's not officially in the docs anywhere but given this philosophy, I usually prefer to use `.toBeVisible` over `.toBeInTheDocument` as it's more idiomatic and is more from the user perspective. As example of why this matters is that an element could be "in the document" but not "be visible." (More info [here](https://github.com/testing-library/jest-dom#tobevisible) on other things it accounts for.)

Can we change these to be `toBeVisible`?

Also a quick tip: The only time I really using `.toBeInTheDocument` is when I `queryBy` something to `.not.toBeInTheDocument`. Or if I want it test that it's in the document but might not be visible to the user.

clairenguyen commented on frontend/__tests__/page.test.tsx:
```diff
@@ -25,4 +47,14 @@ describe('Home', () => {
+    it('the restaurant is saved to the like list', async () => {
+      await userEvent.click(screen.getByRole('button', { name: 'Accept' }))
+
+      const matchList = screen.getByTestId('match-list')
```
Also, related to the same guiding principles mentioned earlier, we generally should also avoid using `data-testid`s when querying for an element as a user would never see or use test ids.  

From the [docs on byTestId](https://testing-library.com/docs/queries/bytestid/):

> In the spirit of [the guiding principles](https://testing-library.com/docs/guiding-principles), it is recommended to use this _only after_ the other queries don't work for your use case. **Using data-testid attributes do not resemble how your software is used and should be avoided if possible.** That said, they are way better than querying based on DOM structure or styling css class names. Learn more about data-testids from the blog post ["Making your UI tests resilient to change"](https://kentcdodds.com/blog/making-your-ui-tests-resilient-to-change)

In your case, I wonder if we can use something else... Although I do see the reason why you chose to use testId here. 🤔 (I'm going to come back to this comment since I have to go right now)

clairenguyen replied:
Didn't have the time to think about alternatives to this. Going to approve the PR and maybe we can come back to this concept at a later point. :)

clairenguyen commented on frontend/src/app/page.tsx:
```diff
@@ -1,12 +1,21 @@
+"use client"
+import LikeList from '@/components/LikeList'
 import CardScreen from '@/components/CardScreen'
+import {useState} from "react";
```
nitpick (non-blocking): noticed some inconsistency with spacing. Since we're not running prettier automatically right now, it would be a good idea to run prettier before making a PR.

clairenguyen commented on frontend/src/components/CardScreen.tsx:
```diff
@@ -3,7 +3,7 @@ import { useEffect, useState } from 'react'
 import CardCarousel from '@/components/CardCarousel'
 
-export default function CardScreen() {
+export default function CardScreen({addLikedRestaurant}) {
```
nitpicking : we should probably add an interface for this component's props

clairenguyen reviewed (changes requested):
Overall, great job! I'm going to a bit more nitpicky as I want us to all elevate our standard of code.  And thus, my main feedback for you is about naming variables. When reading code, clear variable names make it easier for fellow developers (like me) to grasp the meaning and intention quickly. Let's aim for that clarity in our variable naming. Keep it up!

clairenguyen commented on frontend/src/app/page.tsx:
```diff
@@ -1,12 +1,21 @@
+import {useState} from "react";
+import {Restaurant} from "@/types";
 
 export default function Home() {
```
praise: nice this is definitely more readable than it was last week!

clairenguyen commented on frontend/src/app/page.tsx:
```diff
@@ -1,12 +1,21 @@
+import {Restaurant} from "@/types";
 
 export default function Home() {
+  const [likedRestaurants, setLikedRestaurants] = useState<Restaurant[]>([])
```
praise: nice this component is definitely more readable than it was last week!

clairenguyen commented on frontend/src/components/LikeList.tsx:
```diff
@@ -0,0 +1,37 @@
+    likedRestaurants: Restaurant[]
+}
+
+export default function LikeList({ likedRestaurants = [] }: LikeListProps) {
```
praise: renaming this really helps! glad we made that change!

evan-m-jackson committed:
Fix Formatting

- Add/refactor prop interfaces for CardScreen and CardCarousel components
- Run prettier for entire app
- Rename test names/variables for more clarity

clairenguyen reviewed (approved):
Thanks for the changes :) Can you just make to resolve the merge conflicts before you merge? (I'm going to go ahead and approve because i'm going to be out tomorrow and i think github might require the person who requested changes to approve)
